packages/cli/lib/LocalDevManager.js

```diff
@@ -348,32 +367,130 @@ class LocalDevManager {
-    // recently logged warning. We do not want to spam the console with the same message.
-    if (!this.uploadWarnings[warning]) {
-      logger.log();
-      logger.warn(i18n(`${i18nKey}.uploadWarning.header`, { warning }));
-      logger.log(
-        i18n(`${i18nKey}.uploadWarning.stopDev`, {
-          command: uiCommandReference('hs project dev'),
-        })
+    // recently logged warning or the user is actively being prompted or uploading
+    if (
+      this.uploadWarnings[warning] ||
+      this.cancelActivePrompt ||
+      this.isUploading
+    ) {
+      return;
+    }
+
+    logger.log();
+    this.isGithubLinked
+      ? logger.warn(i18n(`${i18nKey}.uploadWarning.githubHeader`, { warning }))
+      : logger.warn(warning);
+
+    if (!this.isGithubLinked) {
+      this.projectUploadPrompt().then(succeeded => {
+        if (!succeeded) {
+          this.projectUploadMessage();
+        } else {
+          this.handleConfigFileChange(filepath);
+        }
+      });
+    } else {
+      this.projectUploadMessage();
+    }
+    this.mostRecentUploadWarning = warning;
+    this.uploadWarnings[warning] = true;
+  }
+
+  async projectUploadPrompt() {
+    uiLine();
+
+    let uploadSuccess = false;
+    const prompt =
+      this.activeApp.type === COMPONENT_TYPES.publicApp
+        ? publicAppUploadPrompt
+        : privateAppUploadPrompt;
+
+    const { cancel, promptPromise } = prompt();
+
+    this.cancelActivePrompt = cancel;
+    const shouldUpload = await promptPromise;
+    this.cancelActivePrompt = null;
+
+    if (shouldUpload) {
+      this.isUploading = true;
+      const { succeeded } = await handleProjectUpload(
+        this.targetProjectAccountId,
+        this.projectConfig,
+        this.projectDir,
+        (...args) => pollProjectBuildAndDeploy(...args, true)
       );
-      if (this.isGithubLinked) {
-        logger.log(i18n(`${i18nKey}.uploadWarning.pushToGithub`));
+      this.isUploading = false;
+
+      if (succeeded) {
+        logger.log(i18n(`${i18nKey}.uploadWarning.prompt.success`));
       } else {
-        logger.log(
+        logger.log(i18n(`${i18nKey}.uploadWarning.prompt.failure`));
+      }
+      uploadSuccess = succeeded;
+    } else {
+      logger.log('');
+      logger.log(i18n(`${i18nKey}.uploadWarning.prompt.decline`));
+    }
+
+    return uploadSuccess;
+  }
+
+  projectUploadMessage() {
+    logger.log(
+      i18n(`${i18nKey}.uploadWarning.stopDev`, {
+        command: uiCommandReference('hs project dev'),
+      })
+    );
+    this.isGithubLinked
+      ? logger.log(i18n(`${i18nKey}.uploadWarning.pushToGithub`))
+      : logger.log(
           i18n(`${i18nKey}.uploadWarning.runUpload`, {
             command: this.getUploadCommand(),
           })
         );
+    logger.log(
+      i18n(`${i18nKey}.uploadWarning.restartDev`, {
+        command: uiCommandReference('hs project dev'),
+      })
+    );
+  }
+
+  updateComponentAtPath(filepath) {
```

Comment:
This re-parses the changed config file and updates `runnableComponents`. This is necessary because the initial parsing of config files happens outside `LocalDevManager`